### ISSUE TITLE
Except admin role, user role can't access and view postgresql custom mutation function. 

### ISSUE THREAD

**Issue by jaekook** [hasura/graphql-engine#8047]
### Version Information

Server Version: 2.1.1
CLI Version (for CLI related issue):

### Environment
Kubernetes

### What is the expected behaviour?

Normal user role can access mutation function by table access permission.
But except admin role can't see and access mutation function. 

Ref.
https://github.com/hasura/graphql-engine/issues/6503

### Keywords

HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS=false

### What is the current behaviour?

Add conf to below config. Then allow permission manually, it's work.

HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS=false

### How to reproduce the issue?

1. With out HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS=false run K8S pod.
2. Add volatile PostgreSQL function to db(using hasura SQL Editor)
3.  Allow permission to role(common_user) Table permission all(Select, update, delete...)
4. Check Hasura Web UI with x-hasura-role="common_user"
5. Not showing new function at mutation list.
6. Uncheck x-hasura-role, admin access we can see the function at mutation

**Comment by PlugFox** [comment]
The same with volatile mutation.
Docker Stack, self-hosted hasura v2.1.1

**Comment by WGTW** [comment]
Using Hasura V2.3.1 I've got the same issue. Custom tracked functions won't have their permissions inferred from the linked tables.

**Comment by BenoitRanque** [comment]
This is not a bug, per the [documentation](https://hasura.io/docs/latest/graphql/core/deployment/graphql-engine-flags/reference/#command-flags):
> When the --infer-function-permissions flag is set to false, a function f, stable, immutable or volatile is only exposed for a role r if there is a permission defined on the function f for the role r, creating a function permission will only be allowed if there is a select permission on the table type. When the --infer-function-permissions flag is set to true or the flag is omitted (defaults to true), the permission of the function is inferred from the select permissions from the target table of the function, only for stable/immutable functions. Volatile functions are not exposed to any of the roles in this case.

Therefore, unless the `HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS` env var is set to `false`, we expect volatile functions to be exposed as mutations and only available to the admin role.

However it seems the actual behavior when an existing function is tracked (as opposed to tracking the function during creation) is slightly different, and the resulting metadata does allow users to set permissions for the function. This may be a bug.

**Comment by Luca8991** [comment]
As I can understand, an SQL function is not exposed as a `mutation` to a user that **doesn't** have an `admin` role, right?

**Comment by BenoitRanque** [comment]
@Luca8991 yes, unless the `HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS` is set to `false`.
If that is the case, the user will be able to explicitly allow the functions on a per-role basis.

**Comment by Luca8991** [comment]
@BenoitRanque it's not working for me I think

**Comment by rogerso** [comment]
I think we are hitting the same bug.
We already have `HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS` set to `false`. However, even with the permission set to allowed through the UI, the function doesn't exist in `mutation_root` for users in the role.

Trying to run the mutation query returns the following error:
```
{"errors":[{"extensions":{"path":"$.selectionSet.the_plpgsql_function","code":"validation-failed"},"message":"field \"the_plpgsql_function\" not found in type: 'mutation_root'"}]}
```

**Comment by BenoitRanque** [comment]
@rogerso can you confirm what server version you are on?

**Comment by rogerso** [comment]
@BenoitRanque `v2.6.2-cloud.1`

**Comment by BenoitRanque** [comment]
@rogerso can you also share your function's definition, and the relevant metadata? Attempting to reproduce this.

**Comment by rogerso** [comment]
Here is a trivial PL/pgSQL function:
```plpgsql
CREATE OR REPLACE FUNCTION users_set_name(hasura_session json, user_id uuid, new_name text) RETURNS users AS $$
DECLARE
  u users;
BEGIN
  PERFORM * FROM users
    WHERE id = user_id AND NOT banned
      AND (id = (hasura_session ->> 'x-hasura-user-id')::uuid OR hasura_session ->> 'x-hasura-role' = 'admin');
  IF NOT FOUND THEN
    RAISE EXCEPTION 'Permission denied';
  END IF;
  UPDATE users SET name = new_name WHERE id = user_id RETURNING * INTO STRICT u;
  RETURN u;
END;
$$ LANGUAGE plpgsql;
```
And the relevant Hasura metadata:
```json
          {
            "function": {
              "schema": "public",
              "name": "users_set_name"
            },
            "configuration": {
              "custom_root_fields": {},
              "session_argument": "hasura_session"
            },
            "permissions": [
              {
                "role": "staff"
              },
              {
                "role": "player"
              }
            ]
          },
```
The function appears in GraphiQL when logged in as admin:
<img width="353" alt="image" src="https://user-images.githubusercontent.com/1010969/169975469-e42d6b6a-0452-4e5d-8c29-32b082c7345e.png">

But not when opened with a JWT with the `staff` or `player` role.

**Comment by harishnair96** [comment]
@rogerso We were able to reproduce this function on `v2.6.2-cloud.1` and was able to confirm that it works as expected. Can you confirm that the table `users` has `select` permissions on the roles `staff` and `player`. According to our [docs](https://hasura.io/docs/latest/graphql/core/databases/postgres/schema/custom-functions/#permissions-for-custom-functions), it is necessary for the returning table of a function to have `select` permissions for the role for the function to show up in the API.

**Comment by rogerso** [comment]
Yes, `player` has `select` permissions, though with the custom check `{"id":{"_eq":"X-Hasura-User-Id"}}` and with access to some columns only.

**Comment by harishnair96** [comment]
@rogerso We were unable to reproduce the issue with the provided information. It might be helpful for us to have more information. Would you mind sharing with us the relevant table metadata, SQL commands used to create the tables, payload of JWT token you are using and also the JWT configuration (you don't have to share the secrets that are part of the JWT configuration)? You can also take a look at the metadata that worked for us so that you can see if there is any difference.
```JSON
{
  "resource_version": 30,
  "metadata": {
    "version": 3,
    "sources": [
      {
        "name": "default",
        "kind": "postgres",
        "tables": [
          {
            "table": {
              "schema": "public",
              "name": "users"
            },
            "select_permissions": [
              {
                "role": "player",
                "permission": {
                  "columns": [
                    "id"
                  ],
                  "filter": {
                    "id": {
                      "_eq": "X-Hasura-User-Id"
                    }
                  }
                }
              },
              {
                "role": "staff",
                "permission": {
                  "columns": [
                    "banned",
                    "name",
                    "id"
                  ],
                  "filter": {}
                }
              }
            ]
          }
        ],
        "functions": [
          {
            "function": {
              "schema": "public",
              "name": "users_set_name"
            },
            "configuration": {
              "custom_root_fields": {},
              "session_argument": "hasura_session"
            },
            "permissions": [
              {
                "role": "player"
              },
              {
                "role": "staff"
              }
            ]
          }
        ],
        "configuration": {
          "connection_info": {
            "use_prepared_statements": false,
            "database_url": {
              "from_env": "PG_DATABASE_URL"
            },
            "isolation_level": "read-committed"
          }
        }
      }
    ]
  }
}
```

**Comment by AymanElarian** [comment]
guys , 
i faced the same issue and fixed by set HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS to false and adjust permation by my self
        - name: HASURA_GRAPHQL_INFER_FUNCTION_PERMISSIONS
          value: "false"
